### PR TITLE
Record: ImprovedParallelResiduals, 1.0758 BPB /   2.7789 nats, -0.0020 BPB / -0.0052 nats vs PR   #1523

### PR DESCRIPTION
**Pull request by msisovic**
# Record: Improved Parallel Residuals

**val_bpb: 1.07578747** (3-seed mean, std 0.0007) | **2.77887078 nats** | **~15.98 MB** | 8xH100 SXM, 600s | Legal TTT

This submission starts from [PR #1523](https://github.com/openai/parameter-golf/pull/1523). Most of the newer submissions moved away from my fuller parallel-residual formulation and settled on a simpler GPT-J-style split-lane decoder. This version keeps the strong parts of that newer baseline and reintroduces the useful parts of my parallel residual implementation.

The key architectural change relative to PR #1523 is in the decoder after the split point. Attention and MLP read from different lanes, but neither sublayer writes back immediately. Instead, both outputs are accumulated into the two lanes together at the end of the block:

```python
next_lane0 = attn_resid * lane0 + attn_post[0] * attn_out + mlp_post[0] * mlp_out
next_lane1 = mlp_resid * lane1 + attn_post[1] * attn_out + mlp_post[1] * mlp_out
```

That keeps the GPT-J-style parallel-in-time update, while restoring the richer learned routing between the attention and MLP lanes. The other important part is that decoder U-Net skips are still written only into `lane0`, which preserves the cheaper and more stable skip path from the newer baseline. Attention reads the mixed `lane0/x0` path, while MLP reads raw `lane1`. Final output uses the mean of the two lanes.

In practice, that is pretty much the only modeling change here versus PR #1523, together with moving `PARALLEL_RESIDUAL_START` from the baseline's `7` to `8`. I ablated that start-layer change separately on top of the plain PR #1523 baseline, without my fuller parallel residual routing changes, and it gave a mild regression on its own. The other notable requirement is that I needed the CUTLASS EVT path to recover the full throughput. In this iteration the CUDA/C++ source is inlined into the training script itself and built against a standard `/opt/cutlass` checkout rather than shipping a separate prebuilt `.so`.

## Results (8xH100 80GB SXM, 600s)

| Seed | Steps | ms/step | Post-EMA BPB | Legal TTT BPB | val_loss (nats) | Artifact |
|------|-------|---------|--------------|----------------|-----------------|----------|
| 1337 | 4,655 | 126.13 | 1.0830 | **1.0751** | 2.7770 | 15,983,095 |
| 2024 | 4,689 | 125.20 | 1.0843 | **1.0765** | 2.7806 | 15,987,382 |
| 42 | 4,696 | 125.04 | 1.0837 | **1.0759** | 2.7790 | 15,982,563 |
| **Mean** | **4680.00** | **125.46** | **1.0837** | **1.07578747** | **2.77887078** | **15984347** |

## Reproducibility

```bash
pip install brotli sentencepiece
git clone https://github.com/NVIDIA/cutlass.git /opt/cutlass
cd /opt/cutlass
git checkout 08185b9c3e90510ee2b656662ed0d53b06d28157
cd -
MATCHED_FINEWEB_REPO_ID=kevclark/parameter-golf python3 data/cached_challenge_fineweb.py --variant sp8192
for SEED in 1337 2024 42; do
    SEED=$SEED TTT_ENABLED=1 HASH_EMBED_ENABLED=1 TTT_LR=0.01 MUON_MOMENTUM=0.97 PARALLEL_RESIDUAL_START=8 GPTQ_RESERVE_SECONDS=13 \
    torchrun --standalone --nproc_per_node=8 train_gpt.py
done
```

### CODE REVIEW

**Comment by mikeapedia** [comment]
Awesome submission! I was just looking through your seed1337 log and I noticed your gptq_reserve_seconds is set to 12.0 in row 21 but the hessian collection took 12.6 seconds (row 204). So you may want to bump the reserve seconds up to 13 just to be safe. I didn't look through the other logs to see if it happened in all 3 runs.

**Comment by msisovic** [comment]
> Awesome submission! I was just looking through your seed1337 log and I noticed your gptq_reserve_seconds is set to 12.0 in row 21 but the hessian collection took 12.6 seconds (row 204). So you may want to bump the reserve seconds up to 13 just to be safe. I didn't look through the other logs to see if it happened in all 3 runs.

Awesome, thanks for noticing! I reran it with 13s reserved, and as expected it didn't noticeably change the score. However, I have noticed that I have on accident ran all three runs with seed 1337, so I have corrected that as well, which was a bit of a hit on the score, but it still clears the bar.

**Comment by MatoTeziTanka** [comment]
### Community Review — Record: ParallelResiduals, 1.0753 BPB

**Compliance:** LOOKS CLEAN — legal score-first-per-chunk TTT (PR #1413 pattern)

PR #1529 ("ImprovedParallelResiduals") implements Test-Time Training via `eval_val_sliding_ttt()` (lines 1324–1476). The TTT logic follows the score-first-per-chunk pattern (PR #1413 pattern) exactly:

**Score phase (lines 1399–1426):**
Each chunk's windows are scored under `torch.no_grad()` before any optimizer step:
```python
with torch.no_grad():
    for bi in range(0, len(my_windows), batch_seqs):
        ...
        logits = compiled_logits(x_batch)
        nll = F.cross_entropy(...)
        loss_sum += scored_nll.sum()
```

**Train phase (lines 1427–1458):**
Training only runs if `not is_last_chunk` — the `is_last_chunk` guard is present at line 1427, preventing optimizer steps on the final scored chunk:
```python
is_last_chunk = ci == num_chunks - 1
if not is_last_chunk and h.ttt_epochs > 0:
    ...
    for _ep in range(h.ttt_epochs):
        optimizer.zero_grad(set_to_none=True)
        loss = base_model(x, y)
        loss.backward()
        optimizer.step()
```

The optimizer trains on the *current* chunk after it has been scored, updating the model for the *next* chunk. This is the canonical legal pattern.

**N-gram check:** No `ctx_hash ^ (target * primes[k])` pattern found anywhere. The only hash embedding is a bigram-style position embedding: `h_idx = (prev_ids * 2039 + input_ids) % h.hash_embed_size` (line 1354), which is `input[t-1] * constant + input[t]` — a legal BigramHash-style positional feature used as an additional learned embedding, not a lookup-table override.

**No SLOT pattern:** No mask+optimize+score-same-region logic detected.

**Architecture note:** The submission adds custom CUTLASS EVT fusion for MLP backward pass (cutlass_evt_fusion) and a Triton TMA fused leaky-relu-squared kernel — these are pure training-efficiency improvements, not eval shortcuts.

**Classification: LEGAL_SCORE_FIRST_TTT_CLEAN**

**Verdict:** LOOKS CLEAN — legal TTT implementation matching the PR #1413 (dexhunter) pattern: each chunk scored under `torch.no_grad()` before `optimizer.step()`, with `is_last_chunk` guard preventing adaptation on the final scored chunk.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks (3-seed validation, under-16MB artifact cap, ≤600s train + ≤600s eval on 8×H100 SXM). TTT implementation follows the legal score-first discipline.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Compliance audit via LLM agent (Sonnet) reviewing full train_gpt.py source, cross-checked against deterministic AST classifier. If this review misread your code, please call it out so I can re-audit manually.*

**Comment by msisovic** [comment]
Thanks for the review @MatoTeziTanka and for catching the `submission.json` inconsistency, it was slightly lagging, showing previously incorrectly captured results, and is now updated to match the correct values from the README.

**Comment by MatoTeziTanka** [comment]
Good to see the alignment fix. The review stands — clean TTT, solid architecture work.

**Comment by simonbissonnette** [comment]
Very nice submission. Just offering a community observation after reading the submitted code:

I agree that I do not see an obvious classic n-gram cache / ctx_hash pattern, and I do not see a classic SLOT-style “mask + optimize + score the same region” pattern either.

That said, I think there are still two possible gray areas worth calling out:

1. Hash embedding during TTT
The hash path does not look like a hard lookup-table override, so I agree it is not the usual explicit n-gram cache pattern. However, it is not just a fixed BigramHash-style feature either. In the submitted script, a fresh hash embedding is created during TTT, zero-initialized, injected into the logits path via h_idx = (prev_ids * 2039 + input_ids) % h.hash_embed_size and then trained during evaluation because hash_emb.weight is added to the TTT parameter set and updated inside the TTT loop.

So my concern is not “this is definitely illegal”, but rather: this looks like a trainable bigram-memory mechanism introduced only at eval time, which feels at least somewhat gray area.

2. CUTLASS sidecar / accounting
The included prebuilt .so is quite large (about 10.6 MB in the submission folder). The README notes the extra source files and gives a stricter size estimate if those are counted, which is helpful. But the challenge README also says that counted code should live in train_gpt.py, so it is not fully obvious to me how OpenAI will want to treat a shipped sidecar binary / extension here.

Again, I am not saying this is invalid. Just that this also seems like a possible gray area for final verification, especially for artifact accounting / self-contained reproducibility.

Overall, impressive submission, thanks for your contribution.

**Comment by msisovic** [comment]
Hi @simonbissonnette thanks for the review, and for pointing those out. Both come from the unaccepted submission that I based my changes on. However, that doesn't mean it shouldn't be compliant. My take on these:

1. This looks legal to me. The issue with previously rejected hash embedding submissions was that the distribution was unscaled. Here, the output of the hash table is added to the logits, yes, but only before the softmax, meaning the output is still a valid distribution.
2. This one may be a real concern, I wasn't aware all code should live in `train_gpt.py`. That being said, it should be relatively easy to inline the custom cutlass kernel into the train_gpt script, so I will do that tomorrow.

**Comment by msisovic** [comment]
I've inlined the custom cutlass kernel from the base PR into my train_gpt script to avoid any compliance issues. The core logic remains unchanged, to keep it fair to submissions that came in the meantime, the results are even a bit worse with the new run, probably down to GPU cluster variance.

**Comment by simonbissonnette** [comment]
> I've inlined the custom cutlass kernel from the base PR into my train_gpt script to avoid any compliance issues. The core logic remains unchanged, to keep it fair to submissions that came in the meantime, the results are even a bit worse with the new run, probably down to GPU cluster variance.

I hope it helps to get it approved. Good luck !

**Review by cocohearts — CHANGES_REQUESTED** [review]
This submission is accepted on substance, but please do a format-only cleanup before we merge it. The record directory is currently `records/track_10min_16mb/2026-04-11_ImrpovedParallelResiduals`; please rename it to fix the typo and use a descriptive standard name such as `records/track_10min_16mb/2026-04-11_ImprovedParallelResiduals_CUTLASS_EVT_LegalTTT`. Please also remove `train_gpt_human.py` unless it is required for the submitted artifact; the mergeable record package should be the scored `train_gpt.py`, `README.md`, `submission.json`, requirements if needed, and the seed logs.

**Review by cocohearts — DISMISSED** [review]
This submission is accepted on substance, but please do a format-only cleanup before we merge it. The record directory is currently `records/track_10min_16mb/2026-04-11_ImrpovedParallelResiduals`; please rename it to fix the typo and use a descriptive standard name such as `records/track_10min_16mb/2026-04-11_ImprovedParallelResiduals_CUTLASS_EVT_LegalTTT`. Please also remove `train_gpt_human.py` unless it is required for the submitted artifact; the mergeable record package should be the scored `train_gpt.py`, `README.md`, `submission.json`, requirements if needed, and the seed logs.

**Comment by msisovic** [comment]
> This submission is accepted on substance, but please do a format-only cleanup before we merge it. The record directory is currently `records/track_10min_16mb/2026-04-11_ImrpovedParallelResiduals`; please rename it to fix the typo and use a descriptive standard name such as `records/track_10min_16mb/2026-04-11_ImprovedParallelResiduals_CUTLASS_EVT_LegalTTT`. Please also remove `train_gpt_human.py` unless it is required for the submitted artifact; the mergeable record package should be the scored `train_gpt.py`, `README.md`, `submission.json`, requirements if needed, and the seed logs.

Thanks for the review, will address this shortly.

**Comment by msisovic** [comment]
@cocohearts Comment addressed, should be ready for merge now.

**Comment by msisovic** [comment]
@cocohearts Ping for this, comment should be addressed.
